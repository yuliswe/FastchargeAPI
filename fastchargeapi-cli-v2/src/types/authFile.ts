export type AuthFileContent = {
  idToken: string;
  refreshToken: string;
  userPK: string;
  email: string;
  issuer: "firebase" | "fastchargeapi" | string;
};

export type RefreshIdTokenResult = {
  idToken: string;
  refreshToken: string;
};

export type VerifyIdTokenResult = {
  email: string;
};
