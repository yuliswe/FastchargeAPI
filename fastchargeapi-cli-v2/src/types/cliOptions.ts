export type CliGlobalOptions = {
  profile?: string;
};

export type CliCommonLoginCommandOptions = {
  profile?: string;
};

export type CliCommonLogoutCommandOptions = {
  profile?: string;
};
