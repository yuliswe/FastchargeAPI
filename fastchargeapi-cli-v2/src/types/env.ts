export type EnvVars = {
  DEV_DOMAIN?: "1";
  TEST?: "1";
  LOCAL_GRAPHQL?: "1";
  LOCAL_PAYMENT?: "1";
  LOCAL_REACT?: "1";
  LOCAL_AUTH?: "1";
  SHOW_AUTH?: "1";
  LOG_REQUESTS?: "1";
};
