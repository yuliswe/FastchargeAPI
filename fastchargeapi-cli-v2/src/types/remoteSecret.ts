export type CliRemoteSercretResult = {
  status: "success" | "canceled";
};
