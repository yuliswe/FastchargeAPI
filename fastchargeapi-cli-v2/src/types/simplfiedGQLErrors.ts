export type SimplifiedGQLErrorExtensions = {};

export type AlreadyExistsSimpleGQLErrorExtensions = SimplifiedGQLErrorExtensions & {
  resource: string;
  query: object;
};

export type BadUserInputSimpleGQLErrorExtensions = SimplifiedGQLErrorExtensions & {
  field: string;
};

export type ImmutableResourceSimpleGQLErrorExtensions = SimplifiedGQLErrorExtensions & {
  resource: string;
};

export type NotFoundSimpleGQLErrorExtensions = SimplifiedGQLErrorExtensions & {
  resource: string;
  query: object;
};

export type PermissionDeniedSimpleGQLErrorExtensions = SimplifiedGQLErrorExtensions & {};

export type TooManyResourcesSimpleGQLErrorExtensions = SimplifiedGQLErrorExtensions & {};

export type ResourceDeletedSimpleGQLErrorExtensions = SimplifiedGQLErrorExtensions & {
  resource: string;
  query: object;
};
