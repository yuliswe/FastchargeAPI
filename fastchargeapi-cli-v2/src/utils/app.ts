import { graphql } from "src/__generated__/gql";
import { GQLClient } from "src/graphqlClient";
import { NotFoundSimpleGQLError } from "src/simplifiedGQLErrors";

export async function getAppPKByPKOrName(args: { client: GQLClient; pkOrName: string }) {
  const { client, pkOrName } = args;
  try {
    const result = await client.mutate({
      mutation: graphql(`
        mutation GetAppPKByName($appName: String!) {
          getAppByName(name: $appName) {
            pk
          }
        }
      `),
      variables: {
        appName: pkOrName,
      },
    });
    return result.data.getAppByName.pk;
  } catch (e) {
    if (e instanceof NotFoundSimpleGQLError) {
      const result = await client.mutate({
        mutation: graphql(`
          mutation GetAppPKById($pk: ID!) {
            getApp(pk: $pk) {
              pk
            }
          }
        `),
        variables: {
          pk: pkOrName,
        },
      });
      return result.data.getApp.pk;
    }
    throw e;
  }
}
