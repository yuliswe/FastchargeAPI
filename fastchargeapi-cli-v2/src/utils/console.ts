import { createInterface } from "readline";

export const print = (message: string) => {
  process.stdout.write(message);
};

export const println = (message: string) => {
  print(message + "\n");
};

export const readline = (prompt: string) =>
  new Promise<string>((resolve) => {
    const rl = createInterface({
      input: process.stdin,
      output: process.stdout,
    });

    rl.question(prompt, (answer) => {
      resolve(answer);
      rl.close();
    });
  });
