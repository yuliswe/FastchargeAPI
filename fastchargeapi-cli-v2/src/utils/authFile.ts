import { promises as fs } from "fs";
import { glob } from "glob";
import { decodeProtectedHeader, importX509, errors as joseErrors, jwtVerify } from "jose";
import fetch from "node-fetch";
import os from "os";
import path from "path";
import { graphql } from "src/__generated__/gql";
import { authServiceHost } from "src/env";
import { getGQLClient } from "src/graphqlClient";
import { tiChecker } from "src/tiChecker";
import { AuthFileContent, RefreshIdTokenResult, VerifyIdTokenResult } from "src/types/authFile";

export async function getGoogleCert(): Promise<{ [hash: string]: string }> {
  const response = await fetch(
    "https://www.googleapis.com/robot/v1/metadata/x509/securetoken@system.gserviceaccount.com"
  );
  return (await response.json()) as { [hash: string]: string };
}

export function getAuthFilePath(profile?: string): string {
  const home = os.homedir();
  if (profile) {
    return path.join(home, ".fastcharge", `auth.${profile}.json`);
  }
  return path.join(home, ".fastcharge", "auth.json");
}

export async function listAuthFiles(): Promise<string[]> {
  const home = os.homedir();
  return await glob([path.join(home, ".fastcharge", "auth.*.json"), path.join(home, ".fastcharge", "auth.json")]);
}

export async function deleteAuthFile(profile?: string) {
  const authFilePath = getAuthFilePath(profile);
  await fs.access(authFilePath);
  await fs.unlink(authFilePath);
}

export async function readAuthFile(profile?: string): Promise<AuthFileContent> {
  const authFilePath = getAuthFilePath(profile);
  const authFileContent = await fs.readFile(authFilePath, "utf-8");
  const auth = tiChecker.AuthFileContent.from(JSON.parse(authFileContent));
  auth.issuer = auth.issuer || "firebase";
  return auth;
}

/**
 * Partially update the auth file context, or create a new auth file with the
 * specified content.
 */
export async function writeToAuthFile(
  profile: string | undefined,
  content: Partial<AuthFileContent>
): Promise<AuthFileContent> {
  const authFilePath = getAuthFilePath(profile);
  await fs.mkdir(path.dirname(authFilePath), { recursive: true });
  let currentAuthFile;
  try {
    currentAuthFile = await readAuthFile(profile);
  } catch {
    currentAuthFile = null;
  }
  // default issuer to firebase
  const newAuthFile = { issuer: "firebase", ...currentAuthFile, ...content };
  await fs.writeFile(authFilePath, JSON.stringify(newAuthFile, null, 2));
  await fs.chmod(authFilePath, 0o600);
  return tiChecker.AuthFileContent.from(newAuthFile);
}

export type VerifyOrRefreshIdTokenResult = {
  refreshed: boolean;
  idToken: string;
  refreshToken: string;
  email: string;
};

export async function refreshIdToken(refreshToken: string): Promise<RefreshIdTokenResult> {
  const resp = await fetch(`${authServiceHost}/refresh-idtoken`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ refreshToken: refreshToken }),
  });

  if (!resp.ok) {
    throw new Error("Unable to refresh id token. " + (await resp.text()));
  }

  return tiChecker.RefreshIdTokenResult.from(await resp.json());
}

export async function verifyIdToken(idToken: string): Promise<VerifyIdTokenResult> {
  const { kid, alg } = decodeProtectedHeader(idToken);

  if (!kid) {
    throw new Error("No kid found in id token.");
  }

  if (!alg) {
    throw new Error("No alg found in id token.");
  }

  const certs = await getGoogleCert();
  const cert = certs[kid];
  if (!cert) {
    throw new Error("No cert found for kid. Google may have rotated the cert and your id token is expired.");
  }

  const key = await importX509(cert, alg);

  try {
    const { payload } = await jwtVerify(idToken, key);
    return tiChecker.VerifyIdTokenResult.from(payload);
  } catch (error) {
    if (error instanceof joseErrors.JWTExpired) {
      throw new Error("Id token expired.");
    }
    throw new Error("Invalid id token. ");
  }
}

export async function verifyOrRefreshIdToken(args: {
  idToken: string;
  refreshToken: string;
  forceRefresh?: boolean;
}): Promise<VerifyOrRefreshIdTokenResult> {
  const { idToken, refreshToken, forceRefresh } = args;
  if (forceRefresh) {
    const newToken = await refreshIdToken(refreshToken);
    const { email } = await verifyIdToken(newToken.idToken);
    return {
      email,
      idToken: newToken.idToken,
      refreshToken: newToken.refreshToken,
      refreshed: true,
    };
  } else {
    try {
      const { email } = await verifyIdToken(idToken);
      return {
        email,
        idToken,
        refreshToken,
        refreshed: false,
      };
    } catch (e) {
      const newToken = await refreshIdToken(refreshToken);
      const { email } = await verifyIdToken(newToken.idToken);
      return {
        email: email,
        idToken: newToken.idToken,
        refreshToken: newToken.refreshToken,
        refreshed: true,
      };
    }
  }
}

/**
 * Gets the id token from the auth file. If the file doesn't exist, returns
 * None. If the id token is issued by Firebase, verifies the token. If the id
 * token is invalid, refreshes it and returns the new id token. If unable to
 * refresh, return None. If the id token is not issued by FastchargeAPI, skips
 * the verification. The new id token and refresh token are written to the auth
 * file.
 */
export async function readOrRefreshAuthFile(args: {
  profile?: string;
  forceRefresh?: boolean;
}): Promise<AuthFileContent | null> {
  const { profile, forceRefresh } = args;
  const auth = await readAuthFile(profile);
  if (auth) {
    if (auth.issuer === "firebase") {
      const user = await verifyOrRefreshIdToken({ ...auth, forceRefresh });
      if (user && user.refreshed) {
        const resp = await getGQLClient(user).query({
          query: graphql(`
            query GetUserPKByEmail($email: Email!) {
              getUserByEmail(email: $email) {
                pk
              }
            }
          `),
          variables: { email: user.email },
        });
        const userPK = resp.data.getUserByEmail.pk;
        return await writeToAuthFile(profile, {
          idToken: user.idToken,
          refreshToken: user.refreshToken,
          email: user.email,
          userPK,
        });
      } else {
        return auth;
      }
    } else if (auth.issuer === "fastchargeapi") {
      return auth;
    } else {
      throw new Error(`Unknown issuer: ${auth.issuer}`);
    }
  }
  return null;
}
