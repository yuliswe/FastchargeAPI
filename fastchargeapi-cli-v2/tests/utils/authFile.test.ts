import fs from "fs/promises";
import os from "os";
import path from "path";
import { AuthFileContent } from "src/types/authFile";
import {
  deleteAuthFile,
  getAuthFilePath,
  getGoogleCert,
  listAuthFiles,
  readAuthFile,
  refreshIdToken,
  verifyIdToken,
  writeToAuthFile,
} from "src/utils/authFile";
import { getExampleAuthFileContent } from "tests/example-data/authFile";
import { getAdminUserCredentials } from "tests/utils";

describe("getAuthFilePath", () => {
  it("returns the correct path", () => {
    expect(getAuthFilePath()).toEqual(path.join(os.homedir(), ".fastcharge", "auth.json"));
    expect(getAuthFilePath("myprofile")).toEqual(path.join(os.homedir(), ".fastcharge", "auth.myprofile.json"));
  });
});

describe("listAuthFiles", () => {
  it("returns the correct files", async () => {
    const file1 = path.join(os.homedir(), ".fastcharge", "auth.json");
    const file2 = path.join(os.homedir(), ".fastcharge", "auth.myprofile.json");
    await fs.mkdir(path.join(os.homedir(), ".fastcharge"));
    await fs.writeFile(file1, "");
    await fs.writeFile(file2, "");
    const files = await listAuthFiles();
    expect(files).toContain(file1);
    expect(files).toContain(file2);
  });

  it("does not throw if .fastcharge directory is not found", async () => {
    await expect(listAuthFiles()).resolves.toEqual([]);
  });
});

describe("getGoogleCert", () => {
  it("returns the correct cert", async () => {
    const cert = await getGoogleCert();
    expect(JSON.stringify(cert)).toContain("-----BEGIN CERTIFICATE-----");
  });
});

describe("deleteAuthFile", () => {
  it("deletes the correct file", async () => {
    const file1 = path.join(os.homedir(), ".fastcharge", "auth.json");
    const file2 = path.join(os.homedir(), ".fastcharge", "auth.myprofile.json");
    await fs.mkdir(path.join(os.homedir(), ".fastcharge"));
    await fs.writeFile(file1, "");
    await fs.writeFile(file2, "");
    await fs.access(file1);
    await fs.access(file2);
    await deleteAuthFile();
    await deleteAuthFile("myprofile");
    await expect(fs.access(file1)).rejects.toThrow();
    await expect(fs.access(file2)).rejects.toThrow();
  });
});

describe("readAuthFile", () => {
  it("returns the correct content", async () => {
    const file1 = path.join(os.homedir(), ".fastcharge", "auth.json");
    const file2 = path.join(os.homedir(), ".fastcharge", "auth.myprofile.json");
    await fs.mkdir(path.join(os.homedir(), ".fastcharge"));
    await fs.writeFile(file1, JSON.stringify(getExampleAuthFileContent(), null, 2));
    await fs.writeFile(file2, JSON.stringify(getExampleAuthFileContent(), null, 2));
    const content1 = await readAuthFile();
    const content2 = await readAuthFile("myprofile");
    expect(content1).toMatchObject(getExampleAuthFileContent());
    expect(content2).toMatchObject(getExampleAuthFileContent());
  });

  it("reject malformed content", async () => {
    const file1 = path.join(os.homedir(), ".fastcharge", "auth.json");
    await fs.mkdir(path.join(os.homedir(), ".fastcharge"));
    await fs.writeFile(file1, "");
    await expect(readAuthFile()).rejects.toThrow();
  });
});

describe("writeToAuthFile", () => {
  it("writes the correct content", async () => {
    const file1 = path.join(os.homedir(), ".fastcharge", "auth.json");
    const file2 = path.join(os.homedir(), ".fastcharge", "auth.myprofile.json");
    await writeToAuthFile(undefined, getExampleAuthFileContent());
    await writeToAuthFile("myprofile", getExampleAuthFileContent());
    const content1 = await fs.readFile(file1, "utf-8");
    const content2 = await fs.readFile(file2, "utf-8");
    expect(JSON.parse(content1)).toMatchObject(getExampleAuthFileContent());
    expect(JSON.parse(content2)).toMatchObject(getExampleAuthFileContent());
  });

  it("throws if the content is incomplete", async () => {
    await expect(writeToAuthFile(undefined, {})).rejects.toMatchSnapshot();
  });
});

describe("verifyIdToken", () => {
  let adminUserCredentials: AuthFileContent;

  beforeAll(async () => {
    adminUserCredentials = await getAdminUserCredentials();
  });

  it("verifies the token", async () => {
    const result = await verifyIdToken(adminUserCredentials.idToken);
    expect(result).toMatchObject({
      email: expect.any(String),
    });
  });
});

describe("refreshIdToken", () => {
  it("refreshes the token", async () => {
    const adminUserCredentials = await getAdminUserCredentials();
    const result = await refreshIdToken(adminUserCredentials.refreshToken);
    expect(result).toMatchObject({
      idToken: expect.any(String),
      refreshToken: expect.any(String),
    });
  });
});
