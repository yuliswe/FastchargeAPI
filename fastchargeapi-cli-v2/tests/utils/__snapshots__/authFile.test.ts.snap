// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`writeToAuthFile throws if the content is incomplete 1`] = `
[Error: Validation failed:
  value.idToken is missing
  value.refreshToken is missing
  value.userPK is missing
Value: {
  "issuer": "firebase"
}]
`;
